wallet/src/main/scala/org/bitcoins/wallet/internal/RescanHandling.scala

```diff
@@ -0,0 +1,157 @@
+package org.bitcoins.wallet.internal
+
+import java.util.concurrent.Executors
+
+import org.bitcoins.core.crypto.DoubleSha256Digest
+import org.bitcoins.core.hd.HDChainType
+import org.bitcoins.core.protocol.script.ScriptPubKey
+import org.bitcoins.core.protocol.{BitcoinAddress, BlockStamp}
+import org.bitcoins.core.util.FutureUtil
+import org.bitcoins.wallet.models.AddressDb
+import org.bitcoins.wallet.{LockedWallet, WalletLogger}
+
+import scala.annotation.tailrec
+import scala.concurrent.{ExecutionContext, Future}
+
+private[wallet] trait RescanHandling extends WalletLogger {
+  self: LockedWallet =>
+
+  val AddressGapLimit = 20
+
+  /////////////////////
+  // Public facing API
+
+  /** @inheritdoc */
+  override def rescanNeutrinoWallet(
+      startOpt: Option[BlockStamp],
+      endOpt: Option[BlockStamp],
+      addressBatchSize: Int): Future[Unit] = {
+
+    logger.info(s"Starting rescanning the wallet.")
+
+    val res = for {
+      _ <- spendingInfoDAO.deleteAll()
+      _ <- addressDAO.deleteAll()
+      _ <- doNeutrinoRescan(startOpt, endOpt, addressBatchSize)
+    } yield ()
+
+    res.onComplete(_ => logger.info("Finished rescanning the wallet"))
+
+    res
+  }
+
+  /** @inheritdoc */
+  override def rescanSPVWallet(): Future[Unit] =
+    Future.failed(new RuntimeException("Rescan not implemented"))
+
+  /////////////////////
+  // Private methods
+
+  private def doNeutrinoRescan(
+      startOpt: Option[BlockStamp],
+      endOpt: Option[BlockStamp],
+      addressBatchSize: Int): Future[Unit] = {
+    for {
+      scriptPubKeys <- generateScriptPubKeys(addressBatchSize)
+      blocks <- matchBlocks(scriptPubKeys, endOpt, startOpt)
+      _ <- downloadAndProcessBlocks(blocks)
+      externalGap <- calcAddressGap(HDChainType.External)
+      changeGap <- calcAddressGap(HDChainType.Change)
+      res <- if (externalGap >= AddressGapLimit && changeGap >= AddressGapLimit)
+        pruneUnusedAddresses()
+      else doNeutrinoRescan(startOpt, endOpt, addressBatchSize)
+    } yield res
+  }
+
+  private def pruneUnusedAddresses(): Future[Unit] = {
+    for {
+      addressDbs <- addressDAO.findAll()
+      _ <- addressDbs.foldLeft(FutureUtil.unit) { (prevF, addressDb) =>
+        for {
+          _ <- prevF
+          spendingInfoDbs <- spendingInfoDAO.findByScriptPubKey(
+            addressDb.scriptPubKey)
+          _ <- if (spendingInfoDbs.isEmpty) addressDAO.delete(addressDb)
+          else FutureUtil.unit
+        } yield ()
+      }
+    } yield ()
+  }
+
+  private def calcAddressGap(chainType: HDChainType): Future[Int] = {
+    for {
```

Comment:
Maybe refactor to this? 

```scala
  private def calcAddressGap(chainType: HDChainType): Future[Int] = {
    val addressDbsF = addressDAO.findAll()
    //make sure all addressDb are of the correct chainType
    //and they are sorted according to their index so we can
    //calculate the gap accurately
    val sortedF = addressDbsF.map {
      _.filter(_.path.chain.chainType == chainType)
        .sortBy(_.path.address.index)
    }
    
    for {
      addressDbs <- sortedF
      addressGap <- addressDbs
        .foldLeft(Future.successful(0)) { (prevNF, addressDb) =>
          for {
            prevN <- prevNF
            spendingInfoDbs <- spendingInfoDAO.findByScriptPubKey(
              addressDb.scriptPubKey)
          } yield {
            if (spendingInfoDbs.isEmpty) prevN + 1 else 0
          }
        }
    } yield {
      logger.debug(s"Address gap: $addressGap")
      addressGap
    }
  }
```

Reply:
Also there might be a potential optmization here if we pass in the `AddressGapLimit` where we can return early... but maybe that is pre-mature. For large wallet's it might make a difference.